src/dp/distributions.rs

```diff
@@ -0,0 +1,323 @@
+/// using the given random number generator for base randomness.
+fn sample_bernoulli_frac<R: Rng + ?Sized>(gamma: &BigURational, rng: &mut R) -> bool {
+    let d = gamma.denom();
+    assert!(!d.is_zero());
+    assert!(gamma <= &BigURational::one());
+
+    // sample uniform biguint in [0,d)
+    let s = rng.gen_range(BigUint::zero()..d.clone());
+
+    s < *gamma.numer()
+}
+
+/// Sample from the Bernoulli(exp(-gamma)) distribution, where $gamma \leq 1$.
+///
+/// `sample_bernoulli_exp1(gamma, rng)` returns numbers distributed as $Bernoulli(exp(-gamma))$.
+/// using the given random number generator for base randomness.
+fn sample_bernoulli_exp1<R: Rng + ?Sized>(gamma: &BigURational, rng: &mut R) -> bool {
+    assert!(!gamma.denom().is_zero());
+    assert!(gamma <= &BigURational::one());
+    let mut k = BigUint::one();
+    loop {
+        if sample_bernoulli_frac(&(gamma / k.clone()), rng) {
+            k += 1u8;
+        } else {
+            return !(k % BigUint::from(2u8)).is_zero();
+        }
+    }
+}
+
+/// Sample from the Bernoulli(exp(-gamma)) distribution.
+///
+/// `sample_bernoulli_exp(gamma, rng)` returns numbers distributed as $Bernoulli(exp(-gamma))$,
+/// using the given random number generator for base randomness.
+fn sample_bernoulli_exp<R: Rng + ?Sized>(gamma: &BigURational, rng: &mut R) -> bool {
+    assert!(!gamma.denom().is_zero());
+    // Sample floor(n/d) independent Bernoulli(exp(-1))
+    // If all are 1, return Bernoulli(exp(-(gamma-floor(gamma))))
+    let mut gamma: BigURational = gamma.clone();
+    while BigURational::one() < gamma {
+        if !sample_bernoulli_exp1(&BigURational::one(), rng) {
+            return false;
+        }
+        gamma -= BigURational::one();
+    }
+    sample_bernoulli_exp1(&gamma, rng)
+}
+
+/// Sample from the geometric distribution with parameter 1 - exp(-gamma) (slow).
+///
+/// `sample_geometric_exp_slow(gamma, rng)` returns numbers distributed according to
+/// $Geometric(1 - exp(-gamma))$, using the given random number generator for base randomness.
+fn sample_geometric_exp_slow<R: Rng + ?Sized>(gamma: &BigURational, rng: &mut R) -> BigUint {
+    assert!(!gamma.denom().is_zero());
+    let mut k = BigUint::zero();
+    loop {
+        if sample_bernoulli_exp(gamma, rng) {
+            k += 1u8;
+        } else {
+            return k;
+        }
+    }
+}
+
+/// Sample from the geometric distribution  with parameter 1 - exp(-gamma) (fast).
+///
+/// `sample_geometric_exp_fast(gamma, rng)` returns numbers distributed according to
+/// $Geometric(1 - exp(-gamma))$, using the given random number generator for base randomness.
+fn sample_geometric_exp_fast<R: Rng + ?Sized>(gamma: &BigURational, rng: &mut R) -> BigUint {
+    let d = gamma.denom();
+    assert!(!d.is_zero());
+    if gamma.is_zero() {
+        return BigUint::zero();
+    }
+
+    // sample uniform biguint in [0,d)
+
```

Comment:
nit: It's a bit odd to have a block comment floating around in space like this (i.e., with no code immediately following it on the next line).